### PR TITLE
Add haiku and jax simple example

### PR DESCRIPTION
**Pull request by nzw0301**
## Motivation

See #2144 

## Description of the changes

I would like to apply optuna to [official MNIST example](https://github.com/deepmind/dm-haiku/blob/master/examples/mnist.py).

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -39,6 +39,7 @@ def get_install_requires() -> List[str]:
         "scipy!=1.4.0",
         "sqlalchemy>=1.1.0",
         "tqdm",
+        "dm-haiku @ git+https://github.com/deepmind/dm-haiku",
```
Comment (crcrpar) [inline setup.py:42]:
```suggestion
```

I think any library required only by example shouldn't be listed here.

**Review by crcrpar — COMMENTED** [review]
Thanks.

Can you
- reduce the size of mnist dataset
- add the pruning feature
like examples/pytorch_simple.py?

**Comment by nzw0301** [comment]
@crcrpar Thank you for your feedback.

1. I have decreased the mini-batch size to make the training faster.
2. I have added the simple `MedianPrunner`.

File: setup.py (hunk 1 above)
Comment (nzw0301) [reply setup.py]: Thanks!

**Comment by hvy** [comment]
Assigned @crcrpar. I don't think it's urgent but let me know if you want it reassigned.

File: examples/haiku_simple.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        new_params = optax.apply_updates(params, updates)
+        return new_params, opt_state
+
+    # Initialize network and optimiser; note we draw an input to get shapes.
+    params = net.init(jax.random.PRNGKey(42), next(train))
+    opt_state = opt.init(params)
+
+    best_test_accuracy = 0.0
+    # Train/eval loop.
+    for step in range(TRAIN_STEPS + 1):
```
Comment (crcrpar) [inline examples/haiku_simple.py:123]:
```suggestion
    for step in range(1, TRAIN_STEPS + 1):
```

I think without start=1, this script update parameters `TRAIN_STEPS + 1`, which sounds not so natural.

File: examples/haiku_simple.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+def load_dataset(
+    split: str,
+    *,
+    is_training: bool,
+    batch_size: int,
+    sample_size: int,
+) -> Generator[Batch, None, None]:
+    """Loads the sub-sampled dataset as a generator of batches."""
+    ds = tfds.load("mnist:3.*.*", split=split).take(sample_size).cache().repeat()
```
Comment (crcrpar) [inline examples/haiku_simple.py:43]: I suspect this is inappropriate because `take` just returns the first `count` samples and so the training dataset is the first 3000 samples, the validation dataset is composed of the first 3000 samples, and the test dataset is the first 1000 samples. This would be a terrible leakage. With `skip` function, I think we can do what I think desirable, i.e., the first 3000 samples for training, following 3000 samples for validation, and the following 1000 samples for testing.

Comment (nzw0301) [reply examples/haiku_simple.py:43]: I would like to clarify the roles of datasets
- `train` is the training dataset.
- `train_eval` is the same as training dataset but large batch size to compute evaluation metric on training dataset faster. This comes from [the original example code](https://github.com/deepmind/dm-haiku/blob/master/examples/mnist.py). 
- `test_eval` is the validation dataset like another example such as [`keras_simple.py`](https://github.com/optuna/optuna/blob/master/examples/keras_simple.py). 

Probably, I should have changed the variable name of `test_eval`. In addition, should I remove `train_eval`?

Comment (crcrpar) [reply examples/haiku_simple.py:43]: Ah, sorry, I missed the "is_training" argument. thanks. `train_eval` being the same as `train` isn't that problematic and kind of common. Thank you for the clarification. :+1:

**Review by crcrpar — APPROVED** [review]

**Review by himkt — COMMENTED** [review]
Thank you for adding the new example! I left some questions, please take a look?

File: setup.py [hunk 4]
```diff
@@ -109,6 +109,9 @@ def get_extras_require() -> Dict[str, List[str]]:
             "dask-ml",
             "botorch ; python_version>'3.6'",
             "fastai",
+            "jax",
+            "optax",
+            "dm-haiku @ git+https://github.com/deepmind/dm-haiku",
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply setup.py:114]: This is because I follow [the haiku's documentation](https://github.com/deepmind/dm-haiku#installation). I'm not sure `pip install dm-haiku` is an official way to install, but [the pip page](https://pypi.org/project/dm-haiku/) is supposed to be maintained by deepmind workers...

File: examples/haiku_simple.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nzw0301) [reply examples/haiku_simple.py:43]: Yes, it shows in my environment due to TensorFlow. Actually, the original example code does as well.

File: examples/haiku_simple.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+N_VALID_SAMPLES = 1000
+
+
+def load_dataset(
+    split: str,
+    *,
+    is_training: bool,
+    batch_size: int,
+    sample_size: int,
+) -> Generator[Batch, None, None]:
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply examples/haiku_simple.py:41]: This function is based on [this function in the example](https://github.com/deepmind/dm-haiku/blob/master/examples/mnist.py#L48).

[the remark this answers is not in the document]
Comment (himkt) [reply examples/haiku_simple.py:41]: I know the annotation comes from the original example but I think if the annotation is not right, we have to correct (and create a PR to dm-haiku). I wanted to know why the method was annotated with this way.

I read the documentation for typing to find `Generator[Batch, None, None]` is correct.
ref. https://docs.python.org/3/library/typing.html#typing.Generator

So, sorry but please ignore this comment.

File: setup.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (himkt) [reply setup.py:114]: Hmm, as you said, `ml-haiku` on PyPI seems to be maintained by deepmind. So I think we can use `dm-haiku` on PyPI, though the documentation recommended to install the software from source. What do you think about it?

File: examples/haiku_simple.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (himkt) [reply examples/haiku_simple.py:43]: Hmm... Can we suppress the warning or fix the root problem? I think this warnings are irritate for users.
(Fixing the root problem may be out of the topic in this PR. I think the simplest workaround is suppressing the message, if it is possible)

File: setup.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nzw0301) [reply setup.py:114]: Either is fine for me, but I think only `pip install dm-haiku` is simpler to maintain for optuna developers. I will update this part. Thanks!

File: examples/haiku_simple.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nzw0301) [reply examples/haiku_simple.py:43]: I couldn't agree more. Definitely, but I'm not familiar with tensorflow, so please give me a time.

File: examples/haiku_simple.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nzw0301) [reply examples/haiku_simple.py:41]: Thank you for investigating it.

**Review by himkt — APPROVED** [review]
Thank you for your contribution! LGTM.

**Comment by nzw0301** [comment]
@crcrpar @himkt Thank both of you for your feedback and suggestions!

**Comment by crcrpar** [comment]
Additional personal comment: Basically, examples are not annotated, but the members of Optuna repository are unfamiliar with Jax and Haiku, so I think it will be better to keep the annotations from haiku repo.

The major difference from the original is removal of exponential moving average of parameters.
